#!/bin/bash

#Amplicon Curation and Strain Assignment Script  - CURRENT AS OF Nov 17, 2021

#So now need to run merging and other amplicon curation, then strain assignment
#Instead of doing QC filtering of specific loci according to controls, going to use database only with relevant strains

#Now hardcoded for each run, but will loop across specified crosses (initally just runB_cross11 as a test)

if [ "$#" -ne 3 ]; then
    echo "You must enter exactly 3 command line arguments: run, cross, and experiment"
    echo "run has format 'runA'"
    echo "File name for cross list. Cross list file is just a plain text file one line per cross: cross1 [break] cross9 etc. "
    echo "Experiment should match the folder generated by demuliplexing script (eg. runA_pairwise_infections or runA_control_infections)"
    exit
fi

#Checking command line parameters:
echo "Checking command line arguments: run, cross list, and experiment"
echo $1
echo $2
echo $3

#Establish base directory where the script resides
BASEDIR=$(dirname "$0")
echo $BASEDIR
INTERMED_DIR="$1""_""$3"
echo $INTERMED_DIR

#Change into folder that has amplicons that requires processing. This folder will be generated by demultiplexing script
cd $BASEDIR/$INTERMED_DIR
pwd
#copy the cross list to make things easier in terms of file paths 
cp $BASEDIR/shared/cross_list_runA_pairwise.txt cross_list.txt

#Initial loop to target only the crosses we need
for cross in `cat cross_list.txt`;
do
  echo "Making a directory for $cross"
  mkdir ${cross}/
  
  cross_number=${cross#cross}
  echo "$cross_number"
  echo "Making custom database for $cross"
  echo "Strains picked:"
  grep ^${cross_number} ../data/cross_data_runA.csv | awk -F "," '{print $2; print $3}'
  
  grep "^${cross_number}," ../data/cross_data_runA.csv | awk -F "," '{print $2; print $3}' | grep -f - ../shared/reference_database.fasta -A1 --no-group-separator > ${cross}/${cross}_reference_database.fasta

  #Now need to do locus aware PEAR merge on all samples in this directory
  for file in ${cross}_sample*_*_R1.fastq;
  do
    prefix=${file%_R1.fastq} #Get file prefix
    echo "Now on file $file"
    echo "$prefix"
    #Merge reads
    pear -f ${prefix}_R1.fastq -r ${prefix}_R2.fastq -o ${cross}/${prefix}_merged.fastq -v 3
  done
done

echo "Custom Databases for each cross done. PEAR R1 and R2 merging done"
echo "Now moving to matching amplicons to our custom databases with usearch, and then tallying matches to asign strains"

#Now do the usearch'ing - HERE AS OF NOV 19, 2021 3:20pm

cd $BASEDIR/$INTERMED_DIR

for cross in `cat cross_list.txt`;
do
#Move into correct directory for cross
cd $BASEDIR/$INTERMED_DIR

#Make directory to hold results
mkdir ${cross}/usearch
mkdir ${cross}/usearch/all

#move into cross directory
cd ${cross}/
echo "Going into $cross folder"
pwd

#Now lookup reads on custom database for each cross (note the loop is going over *all* *.b6 files).
for file1 in cross*_sample*_*_merged.fastq.assembled.fastq;
do
  prefix=${file1%_merged.fastq.assembled.fastq} #Get file prefix
  echo "Now on file $file1"

  #Usearch files
  #Key change for this Oct 20 version is changing the database
  usearch -usearch_global ${file1} -db ${cross}_reference_database.fasta -id 0.98 -blast6out usearch/all/${prefix}_98_merged.b6 -strand both
done
#End the cross loop
done
echo "Amplicon processing done!"


#Now need to add processing and summary

#Now will remove all non-target loci (presumed to be cross amplifications)
cd $BASEDIR/$INTERMED_DIR


for cross in `cat cross_list.txt`;
do
#Move into correct directory for cross
cd $BASEDIR/$INTERMED_DIR

#move into cross directory
cd ${cross}/usearch/all
echo "Going into $cross folder"
pwd

  for merged_file in cross*_sample*_*_98_merged.b6
  do 
    prefix=${merged_file%_98_merged.b6} #Get file prefix
    cross1_sample=${prefix#trimmed_} #Get cross sample combination
    cross1=${cross1_sample%_sample*}
    sample=${cross1_sample#cross*_}
    locus=${sample#sample*_}
    
    echo "$merged_file"
    echo "$prefix"
    echo "$cross1_sample"
    echo "$cross1"
    echo "$sample"
    echo "$locus"

    sed -ni "/${locus}/p" ${merged_file}
    #Double quotes key here, so that shell can interpret the variables
  done

  #Now to summarize the *.b6 results.

  mkdir cd $BASEDIR/$INTERMED_DIR/outputs

  #First let's include all *.b6 files for a given cross_sample set

  for b6_file in ${cross}_sample*_*.b6;
  do
  wc -l $b6_file | awk 'BEGIN { OFS = "\t"; ORS = "\t" } {if($1!="0") print $2, $1}'
  cut -f 2 ${b6_file%} | cut -d , -f 1 | sort | uniq -c | sort -nr | head -n 1 | awk 'BEGIN { OFS = "\t"; ORS = "\n"} {print $1, $2 }'
  done > $BASEDIR/$INTERMED_DIR/outputs/${cross}_pairwise_two_strain_database17_98_locus.txt

#End the cross loop
done
echo "Amplicon processing and strain assignment done!"



